ddtrace/contrib/internal/google_adk/patch.py

```diff
@@ -0,0 +1,256 @@
+        result = None
+        try:
+            agen = wrapped(*args, **kwargs)
+            async for item in agen:
+                yield item
+        except Exception as e:
+            span.set_exc_info(type(e), e, e.__traceback__)
+            raise
+        finally:
+            span.set_tag_str("component", "google-adk")
+            kwargs["tool"] = args[0] if args else kwargs.get("tool")
+            kwargs["tool_args"] = args[1] if len(args) > 1 else kwargs.get("args")
+            kwargs["tool_call_id"] = tool_context.function_call_id
+            integration.llmobs_set_tags(
+                span,
+                args=args,
+                kwargs=kwargs,
+                response=result,
+            )
+
+
+@with_traced_module
+def _traced_code_executor_execute_code(adk, pin, wrapped, instance, args, kwargs):
+    """Trace the execution of code by the agent (sync)."""
+    integration = adk._datadog_integration
+    invocation_context = kwargs.get("invocation_context", args[0] if args else None)
+    provider_name, model_name = extract_provider_and_model_name(agent=getattr(invocation_context, "agent", None))
+
+    # Signature: execute_code(self, invocation_context, code_execution_input)
+    code_input = kwargs.get("code_execution_input", args[1] if len(args) >= 2 and args[1] else None)
+    with integration.trace(
+        pin,
+        "%s.%s" % (instance.__class__.__name__, wrapped.__name__),
+        provider=provider_name,
+        model=model_name,
+        kind="code_execute",
+        submit_to_llmobs=True,
+    ) as span:
+        result = None
+        try:
+            result = wrapped(*args, **kwargs)
+            return result
+        except Exception as e:
+            span.set_exc_info(type(e), e, e.__traceback__)
+            raise
+        finally:
+            span.set_tag_str("component", "google-adk")
+            kwargs["code_input"] = getattr(code_input, "code", None)
+            integration.llmobs_set_tags(
+                span,
+                args=args,
+                kwargs=kwargs,
+                response=result,
+            )
+
+
+def extract_provider_and_model_name(kwargs: Optional[Dict[str, Any]] = None, agent: Any = None) -> Tuple[str, str]:
+    if agent is None:
+        return "google", "google"
+    return agent.model.__class__.__name__, agent.model.model
+
+
+CODE_EXECUTOR_CLASSES = [
+    "BuiltInCodeExecutor",  # make an external llm tool call to use the llms built in code executor
+    "VertexAiCodeExecutor",
+    "UnsafeLocalCodeExecutor",
+    "ContainerCodeExecutor",  # additional package dependendy
+]
+
+# MEMORY_SERVICE_CLASSES = [ # TODO: Add memory tracing
+#     "InMemoryMemoryService",
+#     "VertexAiMemoryBankService",
+#     "VertexAiRagMemoryService",
+# ]
+
+
+def patch():
+    """Patch the `google.adk` library for tracing."""
+
+    if getattr(adk, "_datadog_patch", False):
+        return
+
+    setattr(adk, "_datadog_patch", True)
+    Pin().onto(adk)
```

Comment:
let's see if we can avoid any Pin usage at all in this integration.

it is actively deprecated from the public API and we'll aim to remove all usage internally in 4.0, so could be good to avoid here.

this isn't a blocking comment, just an option to avoid a refactor later.